cmd/upgrade/upgrade.go

```diff
@@ -0,0 +1,242 @@
+)
+
+// SelfUpgrade will return self upgrade command
+func SelfUpgrade(rootCmd *cobra.Command) map[string]cmdCore.CommandEntry {
+	getResourcesFuncs := map[string]cmdCore.CommandEntry{
+		"upgrade": {CmdFunc: selfUpgrade, Aliases: []string{"upgrade"}, ProjectDomainNotRequired: true,
+			Short: upgradeCmdShort,
+			Long:  upgradeCmdLong},
+	}
+	return getResourcesFuncs
+}
+
+func selfUpgrade(ctx context.Context, args []string, cmdCtx cmdCore.CommandContext) error {
+	var source, message string
+	var goos = runtime.GOOS
+	// Get the binary path
+	if err := getExecutable(); err != nil {
+		return err
+	}
+	// Check if it's a rollback
+	if len(args) > 0 {
+		if args[0] == "rollback" {
+			if checkGOOSForRollback(goos) {
+				return nil
+			}
+			// Restore the binary
+			source = backup
+		}
+	} else {
+		latest, err := util.GetLatestVersion("flytectl")
+		if err != nil {
+			return err
+		}
+
+		isGreater, err := util.IsVersionGreaterThan(latest.GetTagName(), stdlibversion.Version)
+		if err != nil {
+			return err
+		}
+		if !isGreater {
+			return nil
+		}
+		if err := upgrade(os.Stdin, latest, ext, goos); err != nil {
+			return err
+		}
+		source = target
+		message = fmt.Sprintf("Successfully updated to version %s", latest.GetTagName())
+	}
+
+	if err := move(source, ext); err != nil {
+		return err
+	}
+
+	if err := os.Chmod(ext, 0111); err != nil {
+		return err
+	}
+
+	if len(message) > 0 {
+		fmt.Println(message)
+	}
+
+	return nil
+}
+
+func upgrade(reader io.Reader, latest *github.RepositoryRelease, ext, goos string) error {
+	// Check if GOOS is windows and darwin
+	if checkGOOSForUpgrade(latest.GetTagName(), goos) {
+		return nil
+	}
+
+	if cmdUtil.AskForConfirmation(fmt.Sprintf("This action will upgrade flytectl version [%s]. Do you want to continue?", latest.GetTagName()), reader) {
+		arch := runtime.GOARCH
+		if arch == "amd64" {
+			arch = "x86_64"
+		} else if arch == "386" {
+			arch = "i386"
+		}
+		for _, v := range latest.Assets {
+			asset := fmt.Sprintf("flytectl_%s_%s.tar.gz", strings.Title(goos), arch)
+			if v.GetName() == asset {
+				response, err := http.Get(*v.BrowserDownloadURL)
+				if err != nil {
+					return err
+				}
+				gzr, err := gzip.NewReader(response.Body)
+				if err != nil {
+					return err
+				}
+				defer gzr.Close()
+
+				tr := tar.NewReader(gzr)
+				for {
+					header, err := tr.Next()
+
+					switch {
+					case err == io.EOF:
+						return nil
+
+					case err != nil:
+						return err
+
+					case header == nil:
+						continue
+					}
+					if header.Name == "flytectl" {
+						f, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+						if err != nil {
+							return err
+						}
+						for {
+							_, err := io.CopyN(f, tr, 1024)
+							if err != nil {
+								if err == io.EOF {
+									break
+								}
+								return err
+							}
+						}
+						// Backup the binary
+						if err := copy(ext, backup); err != nil {
+							return err
+						}
+						break
```

Comment:
Same here can be a part of common functionality to read a tar .
Provide an option to backup as a flag or name of the backup file in the utility